File: modules/buoyant_force_f.m
function F_buoyant = buoyant_force_f(altitude, helium_mass)

%************************************************************************
% Purdue Orbital, Flight Dynamics
%
% Project Name: Ascent/Descent Modeling
%
% Function Name: buoyant_force_f
% File Name: buoyant_force_f.m
%
% Contributors: Aanand Shah, Jack Triglianos
% Date Created: 11/??/2025
% Last Updated: 11/17/2025
%
% Function Description:
%   Computes the buoyant force acting on the balloon using Archimedes'
%   principle. Assumes ideal gas behavior and equilibrium between internal
%   and external pressure and temperature.
%
% References: 
%
% Input variables:
% - altitude: geometric altitude, m, positive
% - helium_mass: mass of helium in the balloon, kg, positive
%
% Output variables:
% - F_buoyant: buoyant force on the balloon, N
%
%************************************************************************

% Compute densities
density_b = balloon_density_f(altitude); % kg/m^3
density_a = air_density_f(altitude); % kg/m^3

% Compute gravitational acceleration
g = gravitationalAcceleration(altitude); % m/s^2

% Compute balloon volume
vol = balloon_volume_f(altitude, helium_mass); % m^3

% Buoyant force (Archimedes)
F_buoyant = (density_a - density_b) * vol * g; % N

File: modules/buoyant_force_f.py
########################################################################
# Purdue Orbital, Flight Dynamics
#
# Project Name: Ascent/Descent Modeling
#
# Function Name: buoyant_force_f
# File Name: buoyant_force_f.py
#
# Contributors: Purdue Orbital Flight Dynamics Team
# Date Created: Unknown
# Last Updated: 02/09/2026
#
# Function Description:
#   Computes buoyant force magnitude (Archimedes principle):
#       F_b = rho_air * V_balloon * g
#
# Notes:
#   - Returns magnitude (always non-negative).
#
# References:
#   None
#
# Input variables:
# - altitude_m: geometric altitude, m, non-negative
# - helium_mass_kg: helium mass in balloon, kg, non-negative
# - atm: atmosphere dict (SI) from modules.atmosphere.atmosphere_m, must include:
#       - rho_kgm3 (kg/m^3)
#
# Output variables:
# - buoyant_force_N: buoyant force magnitude, N, non-negative
#
########################################################################

from __future__ import annotations

from modules.gravity_acceleration_f import gravity_acceleration_f
from modules.volume_balloon_f import volume_balloon_f


def buoyant_force_f(altitude_m: float, helium_mass_kg: float, *, atm: dict) -> float:
    """Return buoyant force magnitude (N)."""
    air_density_kgm3 = float(atm["rho_kgm3"])  # [kg/m^3]
    gravity_mps2 = gravity_acceleration_f(altitude_m)  # [m/s^2]
    volume_m3 = volume_balloon_f(altitude_m, helium_mass_kg, atm=atm)  # [m^3]

    buoyant_force_N = air_density_kgm3 * volume_m3 * gravity_mps2  # [N]
    return float(buoyant_force_N)
